src/classes/drawings/drawing_commands.vala

```diff
@@ -0,0 +1,175 @@
+        // redo commands. Whenever a new DrawingCommand is added, the
+        // redo is cleared (i.e. You can't redo if you've painted something
+        // else. The list is stored in reverse order
+        public List<DrawingCommand?> redo_commands;
+
+        public DrawingCommandList() {
+            clear();
+        }
+
+        public void clear() {
+            this.drawing_commands = new List<DrawingCommand>();
+            this.redo_commands = new List<DrawingCommand>();
+        }
+
+        public void add_line(bool is_eraser,
+                             double x1, double y1, double x2, double y2,
+                             double lwidth,
+                             double red, double green, double blue,
+                             double alpha) {
+            // The new_path which is used for undo and redo is currently
+            // done heuristically by checking if the previous command
+            // is not of the same eraser type or the previous (x2,y2)
+            // is different from the current (x1,y1)
+
+            // After adding a new line you can no longer redo the old
+            // path.
+            this.redo_commands = new List<DrawingCommand>(); // clear
+
+            bool new_path = true;
+            double epsilon = 1e-4; // Less than 0.1 pixel for a 1000x1000 img
+            if (drawing_commands != null) {
+                var last = this.drawing_commands.last().data;
+                if (is_eraser == last.is_eraser
+                    && Math.fabs(last.x2-x1)<epsilon
+                    && Math.fabs(last.y2-y1)<epsilon) {
+                    new_path =false;
+                }
+            }
+
+            var dc = DrawingCommand();
+            dc.new_path = new_path;
+            dc.is_eraser = is_eraser;
+            dc.x1 = x1;
+            dc.y1 = y1;
+            dc.x2 = x2;
+            dc.y2 = y2;
+            dc.lwidth = lwidth;
+            dc.red = red;
+            dc.green = green;
+            dc.blue = blue;
+            dc.alpha = alpha;
+            this.drawing_commands.append(dc);
+        }
+
+        // Paint the drawing commands in the surface
+        public void paint_in_surface(Cairo.ImageSurface surface) {
+            Cairo.Context cr = new Cairo.Context(surface);
+
+            // Clear the surface
+            cr.set_operator(Cairo.Operator.CLEAR);
+            cr.paint();
+
+            // Default settings
+            cr.set_operator(Cairo.Operator.OVER);
+            cr.set_source_rgba(1, 0, 0, 1);
+            cr.set_line_width(5);
+            cr.set_line_cap(Cairo.LineCap.ROUND);
+            cr.set_line_join(Cairo.LineJoin.ROUND);
+
+            // Loop over commands and carry them out
+            int width = surface.get_width();
+            int height = surface.get_height();
+            drawing_commands.foreach((dc) => {
+                if (dc.is_eraser) {
+                    cr.set_operator(Cairo.Operator.CLEAR);
+                }
```

Comment:
change to `} else {`

Reply:
Thanks. I applied all your comment and pushed again.